default/AI/ShipDesignAI.py

```
@@ -0,0 +1,1573 @@
+                print "But the design seems not to exist..."
+                traceback.print_exc()
+                return None
+
+        if verbose:
+            print "Trying to add Design... %s" % name
+        res = fo.issueCreateShipDesignOrder(name, self.description, self.hull.name,
+                                            self.partnames, "", "fighter", False)
+        if res:
+            if verbose:
+                print "Success: Added Design %s, with result %d" % (name, res)
+        else:
+            print "Failure: Tried to add design %s but returned %d, expected 1" % (name, res)
+            return None
+        new_id = _get_design_by_name(name).id
+        if new_id:
+            self.__class__.running_index += 1
+            Cache.map_reference_design_name[reference_name] = name
+            return new_id
+
+    def _class_specific_filter(self, partname_dict):
+        """Add additional filtering to _filter_parts().
+
+        To be implemented in subclasses.
+        """
+        pass
+
+    def optimize_design(self, loc=None, verbose=False):
+        """Try to find the optimimum designs for the shipclass for each planet and add it as gameobject.
+
+        :return: list of (rating,planetID,design_id,cost) tuples, i.e. the best design for each planet.
+        :param loc: int or list of ints (optional) - planet ids where the designs are to be built. Default: All planets.
+        :param verbose: bool - Toggles detailed logging for debugging.
+        """
+        if loc is None:
+            planets = _get_planets_with_shipyard()
+        elif isinstance(loc, int):
+            planets = [loc]
+        elif isinstance(loc, list):
+            planets = loc
+        else:
+            print "ERROR: Invalid loc parameter for optimize_design(). Expected int or list but got", loc
+            return []
+        if self.__class__.__name__ not in Cache.best_designs:
+            Cache.best_designs.update({self.__class__.__name__: {}})
```

sure it returns a value, but it can be silently ignored just as return values often are.  How do you propose to use setitem in a way as simple as can be done here with setdefault?